### PR TITLE
[Perf Experiment] Change FxHash to GxHash

### PR DESCRIPTION
**Pull request by blyxyas**
I've done some experiments comparing hashing libraries that can be found in crates.io, and it while FxHash is faster than the default hasher, Gxhash is still about 10 times faster than fxhash.

This is just with the default implementation, the library permits customization using CPU-specific optimizations. But we can ask the maintainers for that if we were to verify that the optimization is big enough.

### CODE REVIEW

File: blah.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+hola1
```
Comment (lqd) [inline blah.txt:1]: ¡Hola! 👋

**Comment by Kobzol** [comment]
This will be a bit unfair because of the `target-cpu` flag, but let's see.

@bors try @rust-timer queue

**Comment by bjorn3** [comment]
Gxhash seems to be throughput optimized, while rustc-hash is latency optimized. For basically all places where we use FxHasher, latency is much more importanf than throughput as the hash input is relatively small.

**Comment by goffrie** [comment]
Also gxhash [intentionally invokes undefined behaviour](https://ogxd.github.io/articles/unsafe-read-beyond-of-death/) for performance :p

**Comment by RalfJung** [comment]
Yeah that seems to be doing OOB reads and then masking off the OOB parts. That is clear UB. (You need to use "masked load" intrinsics instead that avoid accessing the OOB parts - they exist unstably in intrinsics::simd, not sure if Intel has them as vendor intrinsics.)

**Comment by nnethercote** [comment]
Just FYI, several people have spent a *lot* of time trying to improve on `FxHash`. It's difficult. [This blog post](https://nnethercote.github.io/2021/12/08/a-brutally-effective-hash-function-in-rust.html) has some details.

> Gxhash is still about 10 times faster than fxhash

On what inputs? As @bjorn3 said, rustc's hashing is dominated by tiny inputs, e.g. individual pointers or integers, and fxhash is optimized for those cases, to the point where it's something one one or two *instructions* per input. There's very little room for improvement.

**Comment by blyxyas** [comment]
I've done some more benchmarks, and it seems that, for small benchmarks (3 bytes, 10b, 30b and 100b) `fnv` wins by a pretty great margin.


```
small                fastest       │ slowest       │ median        │ mean          │ samples │ iters
├─ ahash                           │               │               │               │         │
│  ├─ _100_bytes     29.73 ns      │ 5.66 µs       │ 29.73 ns      │ 89.08 ns      │ 100     │ 100
│  ├─ _10_bytes      3.442 ns      │ 3.462 ns      │ 3.451 ns      │ 3.452 ns      │ 100     │ 102400
│  ├─ _30_bytes      4.409 ns      │ 23.86 ns      │ 4.431 ns      │ 4.625 ns      │ 100     │ 51200
│  ╰─ _3_bytes       3.363 ns      │ 3.402 ns      │ 3.383 ns      │ 3.381 ns      │ 100     │ 102400
├─ blake3                          │               │               │               │         │
│  ├─ _100_bytes     144.3 ns      │ 145 ns        │ 144.9 ns      │ 144.8 ns      │ 100     │ 1600
│  ├─ _10_bytes      77.35 ns      │ 77.7 ns       │ 77.7 ns       │ 77.61 ns      │ 100     │ 3200
│  ├─ _30_bytes      77.67 ns      │ 78.01 ns      │ 77.98 ns      │ 77.86 ns      │ 100     │ 3200
│  ╰─ _3_bytes       77.67 ns      │ 78.01 ns      │ 77.7 ns       │ 77.73 ns      │ 100     │ 3200
├─ cityhasher                      │               │               │               │         │
│  ├─ _100_bytes     17.57 ns      │ 17.97 ns      │ 17.58 ns      │ 17.6 ns       │ 100     │ 12800
│  ├─ _10_bytes      5.31 ns       │ 5.37 ns       │ 5.329 ns      │ 5.326 ns      │ 100     │ 51200
│  ├─ _30_bytes      5.525 ns      │ 5.566 ns      │ 5.546 ns      │ 5.541 ns      │ 100     │ 51200
│  ╰─ _3_bytes       5.566 ns      │ 5.859 ns      │ 5.839 ns      │ 5.822 ns      │ 100     │ 51200
├─ default_hasher                  │               │               │               │         │
│  ├─ _100_bytes     27.35 ns      │ 27.6 ns       │ 27.52 ns      │ 27.54 ns      │ 100     │ 12800
│  ├─ _10_bytes      9.05 ns       │ 9.128 ns      │ 9.124 ns      │ 9.11 ns       │ 100     │ 25600
│  ├─ _30_bytes      12.06 ns      │ 12.1 ns       │ 12.1 ns       │ 12.09 ns      │ 100     │ 25600
│  ╰─ _3_bytes       8.382 ns      │ 8.425 ns      │ 8.421 ns      │ 8.422 ns      │ 100     │ 25600
├─ fasthash                        │               │               │               │         │
│  ├─ _100_bytes     17.1 ns       │ 18.67 ns      │ 18.67 ns      │ 18.59 ns      │ 100     │ 12800
│  ├─ _10_bytes      4.526 ns      │ 4.548 ns      │ 4.528 ns      │ 4.532 ns      │ 100     │ 51200
│  ├─ _30_bytes      7.718 ns      │ 7.796 ns      │ 7.757 ns      │ 7.763 ns      │ 100     │ 25600
│  ╰─ _3_bytes       5.056 ns      │ 5.075 ns      │ 5.056 ns      │ 5.057 ns      │ 100     │ 51200
├─ fnv                             │               │               │               │         │
│  ├─ _100_bytes     0.001 ns      │ 0.003 ns      │ 0.003 ns      │ 0.002 ns      │ 100     │ 819200
│  ├─ _10_bytes      0.001 ns      │ 0.003 ns      │ 0.003 ns      │ 0.002 ns      │ 100     │ 819200
│  ├─ _30_bytes      0.001 ns      │ 0.003 ns      │ 0.003 ns      │ 0.002 ns      │ 100     │ 819200
│  ╰─ _3_bytes       0.001 ns      │ 0.471 ns      │ 0.003 ns      │ 0.007 ns      │ 100     │ 819200
├─ fxhash                          │               │               │               │         │
│  ├─ _100_bytes     53.1 ns       │ 53.28 ns      │ 53.26 ns      │ 53.22 ns      │ 100     │ 6400
│  ├─ _10_bytes      0.001 ns      │ 0.003 ns      │ 0.003 ns      │ 0.002 ns      │ 100     │ 819200
│  ├─ _30_bytes      10.88 ns      │ 10.92 ns      │ 10.92 ns      │ 10.91 ns      │ 100     │ 25600
│  ╰─ _3_bytes       0.001 ns      │ 0.003 ns      │ 0.003 ns      │ 0.002 ns      │ 100     │ 819200
├─ gxhash                          │               │               │               │         │
│  ├─ _100_bytes     34.64 ns      │ 35.26 ns      │ 34.81 ns      │ 34.88 ns      │ 100     │ 6400
│  ├─ _10_bytes      9.909 ns      │ 10.18 ns      │ 10.06 ns      │ 10.04 ns      │ 100     │ 25600
│  ├─ _30_bytes      18.13 ns      │ 18.44 ns      │ 18.43 ns      │ 18.4 ns       │ 100     │ 12800
│  ╰─ _3_bytes       10.14 ns      │ 10.26 ns      │ 10.22 ns      │ 10.2 ns       │ 100     │ 25600
├─ highway                         │               │               │               │         │
│  ├─ _100_bytes     41.21 ns      │ 41.54 ns      │ 41.37 ns      │ 41.38 ns      │ 100     │ 6400
│  ├─ _10_bytes      35.73 ns      │ 36.53 ns      │ 36.37 ns      │ 36.4 ns       │ 100     │ 6400
│  ├─ _30_bytes      34.95 ns      │ 35.12 ns      │ 34.96 ns      │ 35.01 ns      │ 100     │ 6400
│  ╰─ _3_bytes       34.95 ns      │ 37.46 ns      │ 37.45 ns      │ 37.2 ns       │ 100     │ 6400
├─ hud_slice_by_8                  │               │               │               │         │
│  ├─ _100_bytes     39.96 ns      │ 136.2 ns      │ 40.12 ns      │ 44.37 ns      │ 100     │ 6400
│  ├─ _10_bytes      6.464 ns      │ 10.26 ns      │ 6.507 ns      │ 6.566 ns      │ 100     │ 25600
│  ├─ _30_bytes      17.18 ns      │ 27.67 ns      │ 22.27 ns      │ 20.99 ns      │ 100     │ 12800
│  ╰─ _3_bytes       6.386 ns      │ 7.972 ns      │ 6.523 ns      │ 6.539 ns      │ 100     │ 51200
├─ meowhash                        │               │               │               │         │
│  ├─ _100_bytes     178.7 ns      │ 181.9 ns      │ 178.8 ns      │ 179 ns        │ 100     │ 1600
│  ├─ _10_bytes      154.9 ns      │ 156.9 ns      │ 155.9 ns      │ 155.9 ns      │ 100     │ 1600
│  ├─ _30_bytes      154.3 ns      │ 156.2 ns      │ 154.4 ns      │ 154.7 ns      │ 100     │ 1600
│  ╰─ _3_bytes       152.4 ns      │ 155.6 ns      │ 152.5 ns      │ 152.8 ns      │ 100     │ 1600
├─ rustc_hash                      │               │               │               │         │
│  ├─ _100_bytes     6.642 ns      │ 6.798 ns      │ 6.778 ns      │ 6.778 ns      │ 100     │ 51200
│  ├─ _10_bytes      2.13 ns       │ 2.141 ns      │ 2.141 ns      │ 2.138 ns      │ 100     │ 102400
│  ├─ _30_bytes      2.795 ns      │ 2.806 ns      │ 2.796 ns      │ 2.797 ns      │ 100     │ 102400
│  ╰─ _3_bytes       2.13 ns       │ 9.254 ns      │ 2.131 ns      │ 2.25 ns       │ 100     │ 102400
├─ wyhash                          │               │               │               │         │
│  ├─ _100_bytes     9.046 ns      │ 9.245 ns      │ 9.202 ns      │ 9.148 ns      │ 100     │ 25600
│  ├─ _10_bytes      5.056 ns      │ 5.077 ns      │ 5.058 ns      │ 5.066 ns      │ 100     │ 51200
│  ├─ _30_bytes      5.78 ns       │ 5.819 ns      │ 5.8 ns        │ 5.797 ns      │ 100     │ 51200
│  ╰─ _3_bytes       4.78 ns       │ 5.058 ns      │ 4.802 ns      │ 4.797 ns      │ 100     │ 51200
├─ xxh3                            │               │               │               │         │
│  ├─ _100_bytes     0.001 ns      │ 0.415 ns      │ 0.003 ns      │ 0.023 ns      │ 100     │ 819200
│  ├─ _10_bytes      0.003 ns      │ 0.348 ns      │ 0.071 ns      │ 0.133 ns      │ 100     │ 409600
│  ├─ _30_bytes      0.067 ns      │ 0.069 ns      │ 0.067 ns      │ 0.068 ns      │ 100     │ 819200
│  ╰─ _3_bytes       0.001 ns      │ 9.697 ns      │ 0.067 ns      │ 0.141 ns      │ 100     │ 819200
╰─ xxhash_rust_xxh3                │               │               │               │         │
   ├─ _100_bytes     24.39 ns      │ 24.47 ns      │ 24.46 ns      │ 24.44 ns      │ 100     │ 12800
   ├─ _10_bytes      31.03 ns      │ 31.12 ns      │ 31.04 ns      │ 31.05 ns      │ 100     │ 12800
   ├─ _30_bytes      29.94 ns      │ 30.03 ns      │ 30.02 ns      │ 29.99 ns      │ 100     │ 12800
   ╰─ _3_bytes       25.17 ns      │ 25.25 ns      │ 25.24 ns      │ 25.25 ns      │ 100     │ 12800
```

So, I'll test it this afternoon, and let's see if the math is mathing

**Comment by Kobzol** [comment]
Some of these numbers looks suspicious, tbh. Not even regarding the differences between hashing algorithms themselves, but more the absolute values - I'm pretty sure you can't hash 3 bytes (much less 100 bytes) in 0.001ns. It might have been optimized out by the compiler entirely, or it might have hit the limits of your timer precision. It might be better to repeat the hashing e.g. 1000x and then measure that whole duration, rather than measuring just the hash of 3 bytes on its own.

But trying e.g. FNV instead of FxHash could be interesting, I'm not sure if such a trivial algo has been tried yet.

**Comment by lqd** [comment]
> But trying e.g. FNV instead of FxHash could be interesting, I'm not sure if such a trivial algo has been tried yet.

https://github.com/rust-lang/rust/pull/37229

**Comment by nnethercote** [comment]
FNV hashes one byte at a time. FxHash can do a whole integer/word at a time.

As I said above, this is *very* well trodden territory.

**Comment by BoxyUwU** [comment]
Closing this since it hasn't had any movement in a month and it seems to just be a perf experiment, feel free to reopen if you are going to work on it again.

**Pull request closed without merging**